R/PosteriorBootstrap.R

```diff
@@ -0,0 +1,486 @@
+  return(rstan::extract(stan_vb)$beta)
+}
+
+#' Stick-breaking depending on a concentration parameter.
+#'
+#' \code{stick_breaking} returns a vector with the breaks of a stick of length 1
+#'
+#' This function implements the stick-breaking process for non-parametric
+#' learning described in section 2 of the supplementary material. The name
+#' "stick-breaking" comes from a stick of unit length that we need to break into
+#' a number of items. This code implements algorithm 2 and the stick-breaking
+#' function calculates the parameter T in algorithm 1, which is the only
+#' difference between the two algorithms. The code uses the Beta distribution as
+#' that distribution is part of the definition of the stick-breaking process.
+#' The function draws from the beta distribution, e.g. \code{b_1}, \code{b_2},
+#' \code{b_3}, ..., and computes the stick breaks as \code{b_1},
+#' \code{(1-b_1)*b_2}, \code{(1-b_1)*(1-b_2)*b_3}, ... . The length remaining in
+#' the stick at each step is \code{1-b_1}, \code{(1-b_1)* (1-b_2)},
+#' \code{(1-b_1)*(1-b_2)*(1-b_3)}, ... so the latter converges to zero.
+#'
+#' @param concentration The parameter \code{c} in the paper (page 3, formula 3),
+#'   which is an effective sample size.
+#' @param min_stick_breaks The minimal number of stick-breaks.
+#' @param threshold The threshold of stick remaining below which the function
+#'   stops looking for more stick-breaks. It corresponds to epsilon in the
+#'   paper, at the bottom of page 5 and in algorithm 2 in page 12.
+#' @return A vector of stick-breaks summing to one.
+#' @examples
+#' stick_breaking(1)
+#' stick_breaking(1, min_stick_breaks = 10)
+#' stick_breaking(1, min_stick_breaks = 10, threshold = 1e-8)
+#'
+#' @export
+stick_breaking <- function(concentration = 1,
+                           min_stick_breaks = 100,
+                           threshold = 1e-8) {
+
+  # This algorithm regenerates all values, which is a waste but is faster than
+  # appending in a for loop, which is slow in R.  The loop is often unnecessary,
+  # as the function is called with min_stick_breaks equal to the the number of
+  # rows in the dataset, which is 1,000, and it only goes once through the
+  # loop. This algorithm also uses some shortcuts in array multiplication to get
+  # the right numbers.
+
+  # The array `stick_remaining` holds the remainder of the stick if we stopped
+  # at that index: 1, 1-beta_draws[1], (1-beta_draws[1])*(1-beta_draws[2]), ...
+  # The algorithm is statistically guaranteed to finish because the last value
+  # of the `stick_remaining` array is a product of numbers between 0 and 1, so
+  # increasing the number of values in that product drives it to zero (even if
+  # we draw new values each time).
+
+  num_stick_breaks <- min_stick_breaks
+  while (TRUE) {
+    # Draw `num_stick_breaks` independent numbers from the
+    # Beta(1, concentration) distribution. All numbers are between 0 and 1.
+    beta_draws <- stats::rbeta(n = num_stick_breaks,
```

Comment:
Is this deterministic - so we're getting the same random numbers at each iteration?